### ISSUE TITLE
--insecure-registry and --registry-mirror don't work together, fallbacks to v1

### ISSUE THREAD

**Issue by ahmetalpbalkan** [docker/docker#15494]
I'm running a local `registry` (as a pull-through cache to Docker Hub) on `localhost:5000` and configuring docker daemon as:

    /usr/bin/docker -d -D -H=unix:// --registry-mirror http://localhost:5000 --insecure-registry localhost:5000

because I have not configured TLS for the registry container yet. In the docker engine logs I can see it falling back to registry protocol v1.

Docker engine logs after a `docker pull` request:

```
DEBU[0008] Calling POST /images/create
INFO[0008] POST /v1.19/images/create?fromImage=ubuntu%3Alatest
DEBU[0008] pinging registry endpoint http://localhost:5000/v0/
DEBU[0008] attempting v2 ping for registry endpoint https://localhost:5000/v2/
DEBU[0008] hostDir: /etc/docker/certs.d/localhost:5000
DEBU[0008] attempting v1 ping for registry endpoint https://localhost:5000/v1/
DEBU[0008] hostDir: /etc/docker/certs.d/localhost:5000
ERRO[0008] Unable to create endpoint for http://localhost:5000/: invalid registry endpoint https://localhost:5000/v0/: unable to ping registry endpoint https://localhost:5000/v0/
v2 ping attempt failed with error: Get https://localhost:5000/v2/: tls: oversized record received with length 20527
 v1 ping attempt failed with error: Get https://localhost:5000/v1/_ping: tls: oversized record received with length 20527. If this private registry supports only HTTP or HTTPS with an unknown CA certificate, please add `--insecure-registry localhost:5000` to the daemon's arguments. In the case of HTTPS, if you have access to the registry's CA certificate, no need for the flag; simply place the CA certificate at /etc/docker/certs.d/localhost:5000/ca.crt
DEBU[0008] pulling image from host "docker.io" with remote name "library/ubuntu"
DEBU[0008] pinging registry endpoint https://index.docker.io/v1/
DEBU[0008] attempting v1 ping for registry endpoint https://index.docker.io/v1/
DEBU[0008] pulling v1 repository with local name "ubuntu"
DEBU[0008] [registry] Calling GET https://index.docker.io/v1/repositories/library/ubuntu/images
DEBU[0008] hostDir: /etc/docker/certs.d/index.docker.io
DEBU[0008] Retrieving the tag list
DEBU[0008] hostDir: /etc/docker/certs.d/registry-1.docker.io
DEBU[0015] Got status code 200 from https://registry-1.docker.io/v1/repositories/library/ubuntu/tags
DEBU[0015] Registering tags
DEBU[0015] Error pulling image (latest) from ubuntu, mirror: http://localhost:5000/v1/, Server error: 404 trying to fetch remote history for 8251da35e7a79dca688682f6da6148a06d358c6f094020844468a782842c2172
DEBU[0015] hostDir: /etc/docker/certs.d/registry-1.docker.io
DEBU[0015] Ancestry: [8251da35e7a79dca688682f6da6148a06d358c6f094020844468a782842c2172 e5855facec0b251bacf22cb50b4f1b4229024eec726061474867b59a3186c8de 5bff21ba54097b77dd18c10e13c8f8db53356813e10fa36b2443ee8964674bd7 6071b4945dcf0886af97cc4782a189cef303fcc43a3d9813aa9fef86e1f7a318]
```

Logs on the `registry` container (built from distribution repo master branch today) –note the `v1`s in there:

```
time="2015-08-11T20:53:11.3115906Z" level=info msg="redis not configured" environment=development instance.id=668586a9-a36e-4b01-97d9-631a2eef7ce4 service=registry version=v2.0.0-559-g1d5b311
time="2015-08-11T20:53:11.3118412Z" level=debug msg="filesystem.Stat(\"/scheduler-state.json\")" environment=development service=registry trace.duration=48.4µs trace.file="/go/src/github.com/docker/distribution/registry/storage/driver/base/base.go" trace.func="github.com/docker/distribution/registry/storage/driver/base.(*Base).Stat" trace.id=701cbf1b-adec-407b-8441-0b1f4696703f trace.line=111 version=v2.0.0-559-g1d5b311
time="2015-08-11T20:53:11.3119568Z" level=info msg="Starting cached object TTL expiration scheduler..." environment=development service=registry version=v2.0.0-559-g1d5b311
time="2015-08-11T20:53:11.3120825Z" level=info msg="scheduler mainloop(): Nothing to do, sleeping..." environment=development service=registry version=v2.0.0-559-g1d5b311
time="2015-08-11T20:53:11.5162329Z" level=info msg="Registry configured as a proxy cache to https://registry-1.docker.io" environment=development instance.id=668586a9-a36e-4b01-97d9-631a2eef7ce4 service=registry version=v2.0.0-559-g1d5b311
time="2015-08-11T20:53:11.51635Z" level=info msg="listening on [::]:5000" environment=development instance.id=668586a9-a36e-4b01-97d9-631a2eef7ce4 service=registry version=v2.0.0-559-g1d5b311
time="2015-08-11T20:53:11.5165305Z" level=info msg="debug server listening localhost:5001"
172.17.42.1 - - [11/Aug/2015:20:53:25 +0000] "GET /v1/images/8251da35e7a79dca688682f6da6148a06d358c6f094020844468a782842c2172/ancestry HTTP/1.1" 404 19 "" "docker/1.7.0 go/go1.4.2 git-commit/0baf609 kernel/3.16.0-31-generic os/linux arch/amd64"
time="2015-08-11T20:55:08.6111025Z" level=info msg="redis not configured" environment=development instance.id=cdf2946c-d5af-48b2-a27c-bf95114f24fa service=registry version=v2.0.0-559-g1d5b311
time="2015-08-11T20:55:08.6113805Z" level=debug msg="filesystem.Stat(\"/scheduler-state.json\")" environment=development service=registry trace.duration=32.3µs trace.file="/go/src/github.com/docker/distribution/registry/storage/driver/base/base.go" trace.func="github.com/docker/distribution/registry/storage/driver/base.(*Base).Stat" trace.id=926d5c0b-315c-4739-85ac-5516eebe5de9 trace.line=111 version=v2.0.0-559-g1d5b311
time="2015-08-11T20:55:08.6115126Z" level=info msg="Starting cached object TTL expiration scheduler..." environment=development service=registry version=v2.0.0-559-g1d5b311
time="2015-08-11T20:55:08.6116338Z" level=info msg="scheduler mainloop(): Nothing to do, sleeping..." environment=development service=registry version=v2.0.0-559-g1d5b311
time="2015-08-11T20:55:08.8400346Z" level=info msg="Registry configured as a proxy cache to https://registry-1.docker.io" environment=development instance.id=cdf2946c-d5af-48b2-a27c-bf95114f24fa service=registry version=v2.0.0-559-g1d5b311
time="2015-08-11T20:55:08.8402105Z" level=info msg="listening on [::]:5000" environment=development instance.id=cdf2946c-d5af-48b2-a27c-bf95114f24fa service=registry version=v2.0.0-559-g1d5b311
time="2015-08-11T20:55:08.8404499Z" level=info msg="debug server listening localhost:5001"
172.17.42.1 - - [11/Aug/2015:20:55:24 +0000] "GET /v1/images/8251da35e7a79dca688682f6da6148a06d358c6f094020844468a782842c2172/ancestry HTTP/1.1" 404 19 "" "docker/1.7.0 go/go1.4.2 git-commit/0baf609 kernel/3.16.0-31-generic os/linux arch/amd64"
```

Any ideas what might be causing this?

cc: @RichardScothern

**Comment by BenHall** [comment]
This might explain https://github.com/docker/distribution/issues/824

In 1.8 I think I had the two working together but I think 1.7 had the issue.

**Comment by thaJeztah** [comment]
@ahmetalpbalkan is this resolved for you in docker 1.8?

**Comment by ahmetalpbalkan** [comment]
@thaJeztah  Yes it's working now in 1.8.1. Thanks.
